Reusable Blocks: Add reusable blocks UI

noisysocks opened the pull request:
Adds the UI necessary for [reusable blocks](https://github.com/WordPress/gutenberg/issues/1516). This includes:

 - PHP code that renders a reusable block in a published post or page
 - The component rendered in the visual editor
 - The sidebar UI used to attach and detach reusable blocks
 - The inserter UI used to insert reusable blocks to the post

Some of this was already reviewed in https://github.com/WordPress/gutenberg/pull/2659.

lamosty commented on editor/effects.js:
```diff
@@ -276,4 +288,78 @@ export default {
+		const { name, type, attributes } = getReusableBlock( getState(), id );
+		const content = serialize( createBlock( type, attributes ) );
+
+		new wp.api.models.ReusableBlocks( { id, name, content } ).save().then(
```
What is the `name` property? The WP REST API [doesn't support](https://developer.wordpress.org/rest-api/reference/posts/#create-a-post) a `name` attribute. Shouldn't it be `title` of the custom post type?

noisysocks replied:
`name` is the name of the reusable block: https://github.com/WordPress/gutenberg/blob/master/lib/class-wp-rest-reusable-blocks-controller.php#L302-L307

I agree that naming that `title` is a better thing to call that, though. I'll put it on my list!

lamosty replied:
Ah, I see, thanks.

Your call. :)

karmatosed commented:
I wanted to test this to give some feedback, however I saw this:

![2017-11-08 at 21 13](https://user-images.githubusercontent.com/253067/32574639-b5d9d83e-c4c9-11e7-9a46-bb74d8377919.png)

Is this a known issue or something I need to do to get it working?

lamosty commented:
@karmatosed:
> I wanted to test this to give some feedback, however I saw this:

After you checkout the branch, try deleting `node_modules`, then `npm install` and `npm run build` again. I had the same issue.

karmatosed commented:
I'm cc'ing in @jasmussen as I think we need to iterate on the design here. Currently we have this:

![2017-11-10 at 20 45](https://user-images.githubusercontent.com/253067/32677916-32ea66ce-c658-11e7-9b30-e10db3bc0ab3.png)

This I am suggesting could be iterated by having some more friendly and informative language.

Here are some suggestions:

![1](https://user-images.githubusercontent.com/253067/32677934-488778fa-c658-11e7-87f9-af69fc4f0933.png)

![2](https://user-images.githubusercontent.com/253067/32677938-4ce008ea-c658-11e7-81f8-a9c32cf9ab9d.png)

jasmussen commented:
I think using the ellipsis menu for making reusable blocks has potential. I'm _slightly_ concerned about putting more stuff in there as we're already looking at a few options there, with a bunch of transformations to come also. But it does feel like a good place for it.

karmatosed commented:
@jasmussen I have that concern myself. The current implementation feels like hiding a good feature though. Maybe we need to try it? Do we have a best guess at how many things we are going to have if everything we want gets in today? I'm thinking of a visual stress test.

jasmussen commented:
I think this may be one of those features whose full potential won't be realized until slightly later on, perhaps post phase-1, where we have more templating features and so on. Given this, perhaps we shouldn't hyper-optimize for the current UI, but instead leave room for us to learn about the feature and then find a better place. To put it differently, there seems to be agreement that it's a slick feature that deserves a better place, but perhaps it's a problem that's best solved in the future?

noisysocks committed:
Add UI for editing a reusable block

Adds the UI which displays when a reusable block is asked for its `edit`
component. This includes a panel which includes UI that indicates
whether a reusable block is loading and lets a user change the name of
the block.

noisysocks committed:
Allow a reusable block to be rendered as HTML

Adds the necessary PHP code for rendering a reusable block as HTML on
the blog's front-end.

noisysocks committed:
Present reusable blocks with a dashed outline

Changes BlockListBlock to display selected reusable blocks with a dashed
outline.

noisysocks committed:
Fix reusable blocks not loading

Fix reusable blocks not loading by looking at the correct path in the
state tree (state.reusableBlocks.data) and by adding a trailing slash to
the /gutenberg/v1/ API versionString.

noisysocks committed:
Add UI for inserting a reusable block

Patches the Inserter to display all reusable blocks in a seperate group,
and to insert a reusable block when one is clicked.

noisysocks commented:
I rebased this PR and fixed some bugs that were introduced in the last rebase! 🙂 

Since `block-inspector/advanced-controls.js` was removed in https://github.com/WordPress/gutenberg/pull/3472, I had to re-implement how the 'Convert to Reusable Block' control is added to the sidebar. It now uses a hook (`hooks/reusable-blocks.js`) which is very similar to how `hooks/custom-class-name.js` works. This does the job, but it feels a little hacky to me since it involves `connect()`ing the `InspectorControls`. 

The E2E tests are failing. A `git bisect` tells me that it's something in https://github.com/WordPress/gutenberg/pull/3378/commits/ce00056313e38c9329b583d215e47c0e5e40de78 that is causing them to break, but I can't for the life of me figure out what it is. My best guess is that Cypress doesn't like it when we use `Slot.Fill`? 😕

karmatosed commented:
@jasmussen I still think it needs to move from where it is currently, I am ok if that for now is the ellipsis.

noisysocks commented:
Thanks for the 👀 @karmatosed! I'll move the controls into the ⋯ menu 🙂

noisysocks commented:
> A git bisect tells me that something in 295e025 is causing them to break, but I can't for the life of me figure out what it is. My best guess is that Cypress doesn't like it when we use Slot.Fill? 😕

For posterity: turns out this was because `<Editable>` was being mounted twice as a result of the `BlockEdit` filter returning a deeply nested array rather than a flat array. 9472e0e was the fix. Thanks @youknowriad for the big hint in figuring this one out!

noisysocks committed:
Add UI for converting blocks to/from a reusable block

Add a block settings menu control which allows users to convert a
regular block to a reusable block and convert a reusable block to a
regular block.

noisysocks commented:
<img width="722" alt="screen shot 2017-11-27 at 13 12 10" src="https://user-images.githubusercontent.com/612155/33247800-dffa13a4-d374-11e7-9af8-d66db66262ca.png">

I moved the button into the block settings menu. @jasmussen—are you able to whip up an icon for this?

jasmussen commented:
Nice work.

While the icon definitely could use some love, I think for now let's go with this dashicon https://developer.wordpress.org/resource/dashicons/#controls-repeat

Reusable blocks is an important interface that I expect we'll hone and polish many times in the future, but for now let's get the ball rolling!

noisysocks commented:
Icon changed!

youknowriad commented on blocks/inspector-controls/button-control/index.js:
```diff
@@ -0,0 +1,23 @@
+/**
```
This component is not used anymore, right?

noisysocks replied:
You're right! I've removed it.

youknowriad commented on editor/components/inserter/group.js:
```diff
@@ -48,7 +48,7 @@ export default class InserterGroup extends Component {
 				key={ block.name }
 				className="editor-inserter__block"
-				onClick={ selectBlock( block.name ) }
+				onClick={ selectBlock( block.name, block.attributes ) }
```
Maybe we should just pass the `block` object as an argument here instead of two arguments.

noisysocks replied:
👌 fixed in 8ced4510795d9e428a2f6d1c2bbb4e2f744da23b.

youknowriad commented on editor/components/inserter/menu.js:
```diff
@@ -103,16 +107,16 @@ export class InserterMenu extends Component {
+	selectBlock( name, attributes ) {
 		return () => {
-			this.props.onSelect( name );
+			this.props.onSelect( name, attributes );
```
Thinking more about this "attributes" argument. Regular blocks already have an `attributes` property but instead of containing `attributes` values, it contains `attributes` definition. I assume this will be called with the `attributes` definitions for these regular blocks which is not what's expected I guess. 

It may not produce bugs at the moment but I think we should fix this ambiguity somehow.

noisysocks replied:
👌 fixed in 8ced4510795d9e428a2f6d1c2bbb4e2f744da23b.

noisysocks committed:
Resolve ambiguity around `onSelect( attributes )`

Distinguish between initialAttributes which are passed into createBlock
and attributes which define which attributes a block type supports.

noisysocks commented:
🤔 looks like Travis CI is getting stuck on a `git clone`...

![screen shot 2017-11-29 at 17 39 47](https://user-images.githubusercontent.com/612155/33361455-5aafbf14-d52c-11e7-88d9-91d1179aaa19.png)
![screen shot 2017-11-29 at 17 39 52](https://user-images.githubusercontent.com/612155/33361459-5bd5e68e-d52c-11e7-86d0-8ad6697f4524.png)

mtias commented:
Looking good. A couple things.

I can search for "reusable block" and insert a broken block:

![image](https://user-images.githubusercontent.com/548849/33438773-075237be-d5ec-11e7-9807-6215983daa3a.png)

My saved block doesn't show up if I search for its name.

![image](https://user-images.githubusercontent.com/548849/33438841-29731a84-d5ec-11e7-83f1-95028add5839.png)

Let's add another tab to the inserter called "Saved" instead of putting saved blocks as a category of "blocks":

![image](https://user-images.githubusercontent.com/548849/33438888-46cbce50-d5ec-11e7-9c4f-17a57f35ec59.png)

mtias commented:
Another point is that the block inspector for the reusable block should let me delete the block from the records (with a warning that it'll stop working if it was used in a post):

![image](https://user-images.githubusercontent.com/548849/33439106-f24e6a62-d5ec-11e7-8d54-9d594bc53dee.png)

mtias commented:
Also for later: creation of global blocks should be handled with proper capabilities (an author would not get access to create reusable blocks, but an editor would, etc). And let this be filterable if someone wants to allow anyone to create reusable blocks.

mtias commented on blocks/library/reusable-block/index.php:
```diff
@@ -0,0 +1,44 @@
+}
+
+register_block_type( 'core/reusable-block', array(
+	'attributes'      => array(
```
Some odd spaces here.

noisysocks replied:
The linter requires that the arrow on line 37 matches up with the arrow on line 43

mtias commented on blocks/library/reusable-block/index.php:
```diff
@@ -0,0 +1,44 @@
+	return gutenberg_render_block( $block );
+}
+
+register_block_type( 'core/reusable-block', array(
```
All the other terms we can more easily change, like internal variables, actions, and functions. But this we'll have to support once we announce.

I am not super comfortable with the name. A couple ideas:

* `core/global-block`
* `core/saved-block` -> to match the inserter tab.
* or maybe this is an exception and is saved like `global-block` without the core suffix. Not to keen on an exception here, but it might warrant it (also of note, we have `core-embed` as a namespace.

For reference, the CPT is now called `wp-block`.

cc @aduth @youknowriad @mcsf @pento for naming thoughts

mtias replied:
I'm not a native speaker, so if `reusable-block` is clearest I'm fine with sticking with it, just want to make sure we are ok with it.

mtias replied:
Thinking a bit more, it seems the most clear and consistent with the `wp-block` post type is to call this `core/block` which maps to `wp:block` and `wp:core/block`. Thoughts?

noisysocks replied:
My only hesitation with _global block_ is that _global_ is a fairly technical term and that using it in the UI might limit how approachable this feature is to a wider audience. 

For technical usages though, I'm happy with using a more precise term. `core/block` matching up with `wp-block` is very appealing.

mcsf replied:
_Saved block_ might speak to users more. Further, though I'm not a huge fan, there are precedents out there to justify _custom blocks_. Note that I can see _reusable blocks_ working, if visual context and copy correctly introduce the concept.

Playing devil's advocate, _saved_ and _custom blocks_ may sound confusing ("Aren't I already making my own block when I add a "Paragraph" and type something in? Isn't it custom? Isn't it being saved?"). Those terms stress a different aspect from _reusable_, with _reusable_ arguably stressing the user-facing value. (Mind you, a similar case can be made against _reusable_: "This Paragraph, or this Author block, aren't they reusable?")

noisysocks replied:
I'm going to go ahead and rename the block identifier to `core/block` so that we can move ahead with this PR. This makes sense to me since, just as `<!-- wp:image -->` inserts an image, `<!-- wp:block -->` inserts a `wp_block`.

Let's move the discussion about the name of the overall feature to a seperate issue—for what it's worth, I asked my wife (a regular WordPress user) what she thought and she said that _custom block_ makes the most sense 🙂

mtias commented on editor/components/block-settings-menu/reusable-block-toggle.js:
```diff
@@ -0,0 +1,49 @@
+			icon="controls-repeat"
+			onClick={ isReusableBlock ? convertToStatic : convertToReusable }
+		>
+			{ isReusableBlock ? __( 'Detach from Reusable Block' ) : __( 'Make into a Reusable Block' ) }
```
Would "Convert to Reusable Block" or "Convert to Global Block" be more concise?

noisysocks replied:
Changed in 619d81f51afbd3bf0fd5a407e2a82a993b7a7051.

mtias commented on lib/blocks.php:
```diff
@@ -131,6 +131,33 @@ function gutenberg_serialize_block( $block ) {
+/**
+ * Renders a single block into a HTML string.
+ *
+ * @since 1.3.0
```
This needs to be updated.

noisysocks replied:
Fixed in b8f96e9b5e7c905d37ba7e392083de1a249cc8a1.

mtias commented on lib/client-assets.php:
```diff
@@ -790,7 +790,7 @@ function gutenberg_editor_scripts_and_styles( $hook ) {
 	$script .= sprintf( 'var editorSettings = %s;', wp_json_encode( $editor_settings ) );
 	$script .= <<<JS
-		window._wpLoadGutenbergEditor = Promise.all( [ wp.api.init(), wp.api.init( { versionString: 'gutenberg/v1' } ) ] ).then( function() {
+		window._wpLoadGutenbergEditor = Promise.all( [ wp.api.init(), wp.api.init( { versionString: 'gutenberg/v1/' } ) ] ).then( function() {
```
curious, what's the reason for this change?

noisysocks replied:
None of the Backbone methods work without the trailing slash. We didn't catch this while testing https://github.com/WordPress/gutenberg/pull/3377 since the unit tests mock the API call and no UI existed for any of this.

youknowriad replied:
Do you think this is still useful with https://github.com/WordPress/gutenberg/pull/3778

noisysocks replied:
Yes—we're not using `withAPIData`.

noisysocks committed:
Change copy of 'Convert to Reusable Block'

noisysocks committed:
Mark gutenberg_render_block as being available @since 1.9.0

noisysocks committed:
Fix inability to search for reusable blocks

Make the Inserter search both static blocks and reusable blocks.

noisysocks committed:
Display reusable blocks in a seperate Inserter tab

Moves reusbale blocks from the Blocks tab to a new Saved tab in the
Inserter.

youknowriad committed:
Reusable Blocks: Fix reusable blocks classname

noisysocks commented on blocks/library/reusable-block/index.js:
```diff
@@ -1,12 +1,160 @@
+		const generatedClassName = hasBlockSupport( blockType, 'className', true ) ?
+			getBlockDefaultClassname( reusableBlock.type ) :
+			null;
+		const className = classnames( generatedClassName, reusableBlockAttributes.className );
```
Clever! Thanks so much @youknowriad 🙏

noisysocks commented:
I rebased this and removed @youknowriad's Autocompleter changes since they were breaking tests. I'll add [those changes](https://gist.github.com/noisysocks/ddfe7d9e58f4e9bb59f5f33acaf1230a) back in a seperate PR.

I think that this is OK to merge. We can iterate on it in seperate (smaller) PRs. I've created some tickets for follow-up work: #3793, #3792, #3791

noisysocks committed:
Rename core/reusable-block to core/block

This way, just as <!-- wp:image --> inserts an image, <!-- wp:block --> inserts a wp_block.

noisysocks commented:
Any last words before I pull the trigger on this?

aduth commented on blocks/library/block/edit-panel/index.js:
```diff
@@ -0,0 +1,68 @@
+					key="save"
+					isPrimary
+					isLarge
+					isIndicatingProgress={ isSaving }
```
This prop does not exist on the `Button` component, and since Button passes through unrecognized props to the underlying DOM node, it results in a warning while editing:

>Warning: React does not recognize the `isIndicatingProgress` prop on a DOM element. If you intentionally want it to appear in the DOM as a custom attribute, spell it as lowercase `isindicatingprogress` instead. If you accidentally passed it from a parent component, remove it from the DOM element.

Should this be the `isBusy` prop added in #3682?

noisysocks replied:
Good catch. It definitely should be `isBusy`. Fixed in a820d1cc43dddd97d44e6e756707dfc6ab56c80b.

aduth commented on blocks/library/block/edit-panel/index.js:
```diff
@@ -0,0 +1,68 @@
+					{ __( 'Detach' ) }
+				</Button>,
+			] }
+			{ ( isEditing || isSaving ) && [
```
Should the editing variation be visible while the name is being saved? In my testing, it flips back to the static text immediately upon hitting Save.

noisysocks replied:
Good catch—that's a bug that must have slipped in during a merge. Fixed in a820d1cc43dddd97d44e6e756707dfc6ab56c80b.

aduth commented on blocks/library/block/index.js:
```diff
@@ -0,0 +1,172 @@
+	}
+}
+
+const ConnectedReusableBlockEdit = connect(
```
I don't have any ideas at the moment and it may have already been mentioned, but we need to work to find a solution where the block can access state without the implicit dependency on editor, either:

- Blocks module having its own separate state
- Merging more of [the non-post-editor-specific-] editor and blocks

Obvious too by the fact that we have equivalent action creators for what we're dispatching here, but unable (unwilling) to import them directly from editor.

noisysocks replied:
Agreed, and I also have no ideas on how to address this right now. 

> Merging more of [the non-post-editor-specific-] editor and blocks

Could you elaborate on this? I'm not quite caught up on my history here. Did we abandon the effort to merge the two modules (https://github.com/WordPress/gutenberg/pull/2795)?

youknowriad replied:
Just an idea: Does it make more sense for this block to be defined in the `editor` module instead?

I also even wonder if the `blocks` module should hold only the blocks API and define all the blocks in the `editor` module

mcsf replied:
> Just an idea: Does it make more sense for this block to be defined in the `editor` module instead?

I think there's a lot of merit to that idea.

> I also even wonder if the `blocks` module should hold only the blocks API and define all the blocks in the `editor` module

That one I'm not too convinced of right now. I mean, I could picture using `editor/library`, but the current situation of having the library sit next to the API makes sense to me, as the core blocks are the canonical blocks: in that sense, they highly complement the API (both code and docs) by providing numerous well tested use cases.

aduth commented on blocks/library/block/index.js:
```diff
@@ -0,0 +1,172 @@
+	category: 'reusable-blocks',
+	isPrivate: true,
+
+	attributes: {
```
I don't think you should need to define this on the client, since it should be bootstrapped from the server definition.

See #2529

noisysocks replied:
I'm hesitant to do this because removing the client-side definition means our test fixtures aren't able to test parsing a `<!-- wp:block ref="358b59ee-bab3-4d6f-8445-e8c6971a5605" /-->`.

aduth commented on components/spinner/index.js:
```diff
@@ -1,3 +1,10 @@
+import classnames from 'classnames';
 
-export default () => spinner;
+function Spinner( { className } ) {
```
You're not / no longer taking advantage of these enhancements? Should we omit them respecting our good friend [YAGNI](https://en.wikipedia.org/wiki/You_aren%27t_gonna_need_it)?

noisysocks replied:
👍 I've removed these changes.

aduth commented on editor/components/block-list/block.js:
```diff
@@ -350,11 +350,13 @@ class BlockListBlock extends Component {
 		const { isHovered, isSelected, isMultiSelected, isFirstMultiSelected, focus } = this.props;
 		const showUI = isSelected && ( ! this.props.isTyping || ( focus && focus.collapsed === false ) );
 		const { error } = this.state;
+		const isReusableBlock = blockType.name === 'core/block';
```
Elsewhere we've limited our awareness to specific block types via abstraction, e.g. `getDefaultBlockName`, `getUnknownTypeHandlerName`. Should we need to consider that here? Maybe not the case we'd ever expect it to change.

noisysocks replied:
Couldn't hurt! I moved this logic out to `isReusableBlock` in 95a28a37adea3afeb44fb2b9c068c9e1fc61fb31.

aduth commented on editor/components/block-settings-menu/reusable-block-toggle.js:
```diff
@@ -0,0 +1,49 @@
+import { convertBlockToStatic, convertBlockToReusable } from '../../actions';
+
+export function ReusableBlockToggle( { block, convertToStatic, convertToReusable } ) {
+	const isReusableBlock = block.name === 'core/block';
```
Thinking purely from the [presentational / container distinction](https://medium.com/@dan_abramov/smart-and-dumb-components-7ca2f9a7c7d0), I might expect this variable to be passed as the prop, determined in a selector or `mapStateToProps`, particularly since we're not otherwise using the `block` prop.

noisysocks replied:
👍 I've moved it in 95a28a37adea3afeb44fb2b9c068c9e1fc61fb31.

aduth commented on editor/components/inserter/menu.js:
```diff
@@ -71,6 +71,10 @@ export class InserterMenu extends Component {
 	}
 
+	componentDidMount() {
+		this.props.fetchReusableBlocks();
```
Noting that this can become quite noisy for AJAX requests: Every time the user opens an inserter.

Personally I'm okay with this, since it ensures that the data is kept fresh, but worth keeping in mind.

noisysocks replied:
For now I think this is OK. If it's a problem then we can limit it to fetch once when the inserter is opened for the first time.

aduth reviewed (commented):
Most comments aren't critical. This works great and is _very_ cool to use and see 👍

noisysocks committed:
Fix Reusable Block saving state

Ensure that the Save button indicates that the block is being saved.

noisysocks committed:
DRY up block.name === 'core/block' logic

Move block.name === 'core/block' logic to a new isReusableBlock
function.

mtias commented:
@noisysocks Before we go ahead with exposing this at wide. Can you remind me why we are not using the ID provided by WP for the post object in `wp_block` like `public 'ID' => int 2359` instead of the generated `ref`? It makes a lot more sense for me to see it as `<!-- wp:block id="2359" /-->` which is an internally consistent concept of WordPress.

youknowriad committed:
Reusable Blocks: Avoiding using a generated uid and rely on the post_id instead

mcsf commented on editor/reducer.js:
```diff
@@ -731,6 +754,22 @@ export const reusableBlocks = combineReducers( {
+				return {
+					...omit( state, id ),
+					[ updatedId ]: {
+						...omit( state[ id ], [ 'id', 'isTemporary' ] ),
```
Totally minor, but omitting `'id'` isn't strictly necessary.

mcsf reviewed (approved):
Per Slack, this looks great. The recent change to server-issued IDs is quite welcome.

mtias commented:
Great work here everyone and specially @noisysocks. This is a great feature.

noisysocks commented:
> Also for later: creation of global blocks should be handled with proper capabilities (an author would not get access to create reusable blocks, but an editor would, etc). And let this be filterable if someone wants to allow anyone to create reusable blocks.

@mtias — could you please elaborate on this so that I can create an issue? Right now, any user with the `edit_posts` capability (authors, contributors and editors) can view and edit Reusable Blocks. Are you suggesting that we limit editing and creating Reusable Blocks to _just_ editors?

mtias commented:
> Are you suggesting that we limit editing and creating Reusable Blocks to just editors? Right now, any user with the edit_posts capability (authors, contributors and editors) can view and edit Reusable Blocks.

We need to let authors insert them, but probably not edit them by default. Not sure if we should let authors create them, or only editors. Furthermore, this is something that should probably be filterable and part of general capabilities so people can map. Example: `edit_blocks`, `insert_blocks`, `delete_blocks`. Something more broad to think about. cc @pento

tkes commented:
@mtias ... something more broad.

There indeed needs to be a complete capabilities- and author-concept for reusable blocks.

A current users with a role below "editor" should only be able to see/create/edit/delete **his/her own** reusable blocks.

Users with/above "editor" role should be able to see/create/edit/delete **any** reusable block and also edit its author.

The concept should also provide a way to configure the visibility of reusable blocks for each user and/or post_type, not sure if via capability or filter.

Example for no visibility limit requirement: Simple one-user blog.

Example for visibility limit requirement: Multi author website with 15-20 authors and everybody creates a few reusable blocks which would then be visible for everybody, the gui would be a mess in no time...
